File: Graph/CountUnreachablePairs/count.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int countNodeInConnComp(vector<bool>& visited,
                            vector<vector<int>>& adjList,
                            int node) {

        queue<int> qu;
        qu.push(node);
        visited[node] = true;

        int totalNodes = 0;

        while (!qu.empty()) {
            int currNode = qu.front();
            qu.pop();

            totalNodes++;

            for (int neighbour : adjList[currNode]) {
                if (!visited[neighbour]) {
                    visited[neighbour] = true;
                    qu.push(neighbour);
                }
            }
        }

        return totalNodes;
    }

    long long countPairs(int n, vector<vector<int>>& edges) {

        vector<vector<int>> adjList(n);
        vector<bool> visited(n, false);

        for (auto& edge : edges) {
            adjList[edge[0]].push_back(edge[1]);
            adjList[edge[1]].push_back(edge[0]);
        }

        vector<int> totalNodeInComps;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                totalNodeInComps.push_back(
                    countNodeInConnComp(visited, adjList, i)
                );
            }
        }

        int totalComps = totalNodeInComps.size();

        if (totalComps == 1)
            return 0;

        vector<long long> suffixSum(totalComps, 0);

        long long totalComp = 0;

        for (int i = totalComps - 1; i >= 0; i--) {
            suffixSum[i] = totalComp;
            totalComp += totalNodeInComps[i];
        }

        long long totalUnreachablePairs = 0;

        for (int i = 0; i < totalComps; i++) {
            totalUnreachablePairs +=
                suffixSum[i] * totalNodeInComps[i];
        }

        return totalUnreachablePairs;
    }
};

File: Graph/CountUnreachablePairs/count.py
from collections import deque
from typing import List

class Solution:
    def countNodeInConnComp(self, visited, adjList, node):
        qu = deque()
        qu.append(node)
        totalNodes = 0
        visited[node] = True
        while len(qu):
            currNode = qu.popleft()
            totalNodes += 1
            for neighbour in adjList[currNode]:
                if not visited[neighbour]:
                    qu.append(neighbour)
                    visited[neighbour] = True
        return totalNodes

    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        adjList = [[] for _ in range(n)]
        visited = [False] * n
        totalEdges = len(edges)
        for i in range(totalEdges):
            adjList[edges[i][0]].append(edges[i][1])
            adjList[edges[i][1]].append(edges[i][0])
        totalNodeInComps = []
        for i in range(n):
            if not visited[i]: 
                totalNodeInComps.append(self.countNodeInConnComp(visited,adjList,i))
        totalComps = len(totalNodeInComps)
        if totalComps==1: return 0
        suffixSum = [0] * totalComps
        totalComp = 0
        for i in range(totalComps-1,-1,-1):
            suffixSum[i] = totalComp
            totalComp += totalNodeInComps[i]
        totalUnreachablePairs = 0
        for i in range(totalComps):
            totalUnreachablePairs += (suffixSum[i]*totalNodeInComps[i])
        return totalUnreachablePairs